system/lib/emmalloc.c

```diff
@@ -352,38 +352,36 @@ static void dump_memory_regions()
 {
   ASSERT_MALLOC_IS_ACQUIRED();
   RootRegion *root = listOfAllRegions;
-  MAIN_THREAD_ASYNC_EM_ASM(console.log('All memory regions:'));
+  _emscripten_out("All memory regions:");
   while(root)
   {
     Region *r = (Region*)root;
     assert(debug_region_is_consistent(r));
     uint8_t *lastRegionEnd = root->endPtr;
-    MAIN_THREAD_ASYNC_EM_ASM(console.log('Region block 0x'+($0>>>0).toString(16)+' - 0x'+($1>>>0).toString(16)+ ' ('+($2>>>0)+' bytes):'),
-      r, lastRegionEnd, lastRegionEnd-(uint8_t*)r);
+    _emscripten_outf("Region block %p - %p ' (%zu bytes):", r, lastRegionEnd, lastRegionEnd-(uint8_t*)r);
     while((uint8_t*)r < lastRegionEnd)
     {
-      MAIN_THREAD_ASYNC_EM_ASM(console.log('Region 0x'+($0>>>0).toString(16)+', size: '+($1>>>0)+' ('+($2?"used":"--FREE--")+')'),
-        r, r->size, region_ceiling_size(r) == r->size);
+      _emscripten_outf("Region %p, size: %zu (%s)", r, r->size, region_ceiling_size(r) == r->size ? "used" : "--FREE--");
 
       assert(debug_region_is_consistent(r));
-      uint32_t sizeFromCeiling = size_of_region_from_ceiling(r);
+      size_t sizeFromCeiling = size_of_region_from_ceiling(r);
       if (sizeFromCeiling != r->size)
-        MAIN_THREAD_ASYNC_EM_ASM(console.log('Corrupt region! Size marker at the end of the region does not match: '+($0>>>0)), sizeFromCeiling);
+        _emscripten_outf("Corrupt region! Size marker at the end of the region does not match: %zu", sizeFromCeiling);
```

Comment:
Noo :( This will pull in printf() facilities into the build, which will explode code size on small programs. Especially for debug builds, which are exactly the ones that are most often being wasm-dissed, so there will be a increase to code size on minimal repros.

The reason for using MAIN_THREAD_ASYNC_EM_ASM() was that web browsers do not provide any guarantees that calling console.log() or console.error() are lock-free, i.e. won't internally hold a lock from the main browser thread, and won't be proxied by the browser to the main thread.

Since these print functions are called potentially from a worker thread, and from inside an emmalloc acquire lock, calling console.log() there could make the main browser thread and the worker thread deadlock.

I would prefer the regular `EM_ASM()` or `EM_JS()` for logging, since it has an absolutely tiny code size foot print. This also produces a static guarantee that these logging functions won't allocate - I don't know if printf() in libc is required to have that guarantee.

Reply:
The idea being having the _emscripten_out and _emscripten_err C functions is so that low level libraries can log stuff without having to roll their own EM_ASM/EM_JS mechanism.

_emscripten_out and _emscripten_err don't necessarily map to console.error and console.log.    For example under node we map them to `stdout.write` and `stderr.write` to avoid the posiblity of loosing message that you mention.  Perhaps on the browser side we should proxy them to main the thread synchronously.

In other words if `_emscripten_out` and `_emscripten_err` need fixing to be relaible then we should fix them, rather than avoiding them in this one place.

Reply:
Regarding the code side issue that having native printf adds, I suppose we could instead do the formatting on the JS side?  I'll take a look at that possibility.